typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,916 @@
+  .with({ type: 'ok', data: { type: 'img', src: select() } }, (src) => `<img src=${src} />`)
+  .exhaustive();
+```
+
+### Immutability
+
+TypeScript has two base primitives to work with immutability: const and readonly.
+
+The first one is used to prevent variable reference change.
+
+```typescript
+const a = 1;
+a = 2; // Cannot assign to 'a' because it is a constant.
+```
+
+And the second one for properties immutability.
+
+```typescript
+interface A {
+  readonly x: number;
+}
+const a: A = { x: 1 };
+a.x = 12; // Cannot assign to 'x' because it is a read-only property.
+```
+
+Nevertheless, you can easily see problems here.
+Const and readonly only block the reference change, but make nothing about values.
+With `const a = [1, 2, 3]` or `readonly x: number[]` you still will be able to change the content of array.
+Let's introduce more cases and how you can deal with them in TypeScript.
+
+
+The first one is `Readonly` type. With it you can mark all properties as `readonly`.
+```typescript
+type A = {
+  x: number;
+  y: number;
+};
+
+type ImmutableA = Readonly<A>; 
+const a: ImmutableA = { x: 1, y: 2 };
+a.x = 2; // Cannot assign to 'x' because it is a read-only property.
+a.y = 1; // Cannot assign to 'y' because it is a read-only property.
+```
+
+The second one is `ReadonlyArray` which make all array elements as `readonly`.
+
+```typescript
+const arr: ReadonlyArray<number> = [1, 2, 3];
+arr[0] = 4; // Index signature in type 'readonly number[]' only permits reading.
+```
+
+You can also make `readonly` dictionary for example.
+```typescript
+interface A {
+    readonly [x: string]: number;
+}
+
+const a: A = { x: 1, y: 2 };
+a.x = 2; // Index signature in type 'A' only permits reading.
+a.y = 1; // Index signature in type 'A' only permits reading.
+```
+
+Also be careful passing object with readonly fields to function as it may be changed inside the function.
+It is happening because of information loss.
+Object with the type which include `readonly` is a subtype of the type without it and function argument with more common type know nothing about its subtype.
+As a result, you should always write proper types of arguments.
+
+```typescript
+const a: { readonly x: number } = { x: 1 };
+
+const changeA = (arg: { x: number }): void => {
+  const argC = arg;
+  argC.x = 2;
+}
+
+changeA(a);
+a.x; // 2
+```
+
+Now you get more flexible way to make things readonly.
+Although now you can get immutable objects usually you don't need them as is.
+You want to make some actions on them and make new immutable objects.
+But it is not so easy.
+You will need to construct new objects by hands, you can also use `Writable` utility type, but it still is not really simple to copy and construct new objects.
+
+Let's introduce some libraries which may help you deal with this issue.
+The first one and the most powerful is [immer](https://immerjs.github.io/immer/).
+It is used for immutable state flow.
+Immutable data is not copied but shared in memory.
```

Comment:
Is this a statement about all immutable data or about what immer lets you to do?